README.md
termcolor_dg
============

📦 [pypi/termcolor_dg](https://pypi.python.org/pypi/termcolor_dg) |
📑 [ANSI escape code](https://en.wikipedia.org/wiki/ANSI_escape_code) |
📑 [Color codes cheatsheet](https://delameter.github.io/termcolor)

ANSI Color formatting for terminal output and log coloring. Supports 16 colors, 256 colors and 24-bit color modes.

Python 2 support is present for legacy projects, and because it is not too much work and I have to use it for now.


Example
-------

```python
>>> from termcolor_dg import colored, cprint
>>>
>>> print(colored('Hello, World!', 'light_red', 'on_blue'))
>>> cprint('Hello, World!', 'blue', 'on_light_red', ['reverse', 'blink'])
>>>
>>> print_red_on_cyan = lambda x: cprint(x, 'red', 'on_cyan')
>>> print_red_on_cyan('Hello, World!')
>>>
>>> for i in range(10):
>>>     cprint(str(i), 'magenta', end=' ' if i != 9 else '\n')
>>>
>>> import sys
>>> cprint("Attention!", 196, attrs=('bold'), file=sys.stderr)  # 256 color mode
>>> cprint("Attention!", (255, 0, 0), attrs=('bold'), file=sys.stderr)  # 24 bit color mode
>>>
>>> import logging
>>> from termcolor_dg import logging_basic_color_config
>>> logging_basic_color_config()
>>> logging.log(logging.INFO, 'test')
```


Colors demo screenshot (`python -m termcolor_dg`):
--------------------------------------------------

![colors.png](colors.png "Colors demo")


Colored logs demo screenshot (`python -m termcolor_dg logs`):
-------------------------------------------------------------

![color_logs.png](color_logs.png "Colorized logs demo")


Environment variables
---------------------

- **ANSI_COLORS_FORCE**
  if set to anything, even empty string, color escape sequences will be added.

- **ANSI_COLORS_DISABLED**
  if set to anything no coloring will be performed, overrides **ANSI_COLORS_FORCE**.

- **NO_COLOR**
  if set to anything no coloring will be performed, overrides **ANSI_COLORS_FORCE**,
  see https://no-color.org/.

If none of the environment variables is set the colors are used
only if the ``stdout`` is attached to a terminal: ``sys.stdout.isatty()``.


Text properties
---------------

| Text colors   | Text highlights | Alt Text highlights | Attributes |
|---------------|-----------------|---------------------|------------|
| black         | black           | on_black            | bold       |
| red           | red             | on_red              | dark       |
| green         | green           | on_green            | underline  |
| yellow        | yellow          | on_yellow           | blink      |
| blue          | blue            | on_blue             | reverse    |
| magenta       | magenta         | on_magenta          | concealed  |
| cyan          | cyan            | on_cyan             |            |
| light_grey    | light_grey      | on_light_grey       |            |
| dark_grey     | dark_grey       | on_dark_grey        |            |
| light_red     | light_red       | on_light_red        |            |
| light_green   | light_green     | on_light_green      |            |
| light_yellow  | light_yellow    | on_light_yellow     |            |
| light_blue    | light_blue      | on_light_blue       |            |
| light_magenta | light_magenta   | on_light_magenta    |            |
| light_cyan    | light_cyan      | on_light_cyan       |            |
| white         | white           | on_white            |            |


Functions
---------

### Adding ANSI colors

**always_colored(text, color=None, on_color=None, attrs=None, reset=True)**
:   Returns the text with ANSI color code in front and ANSI color reset after. Arguments:
:   - **text** is the text to add color to
:   - **color** and **on_color** define the character and background color. Each can be a name for 16 color mode, number for the 256 color variant and list/tuple of R, G and B for the 24 bit color support. The following all define light red: **'light_red'**, **9**, **196** and **(255, 0, 0)**.
:   - **attrs** attributes, single attribute name, list or tuple of attributes.
:   - **reset** if set to **False** will suppress adding of the reset ANSI sequence at the end. Useful if you are joining another colored string, only the last reset is really needed.

**colored(text, color=None, on_color=None, attrs=None, reset=True)**
:   same as always_colored but checks if the app is running on a terminal and if the colors have been forced or disabled. The boolean **termcolor_dg.DISABLED** is the variable checked.


### Utility functions

**cprint(text='', color=None, on_color=None, attrs=None, \*\*kwargs)**
:   same as colored, but prints the resulting string instead of returning it.

**rainbow_color(n, steps, nmax=255)**
:   calculates the color at step **n** needed to produce a rainbow in **steps** steps with intensity up to **nmax**. Ex: **rainbow_color(0, 120)** returns **(255, 0, 0)**, **rainbow_color(30, 120)** returns **(128, 255, 0)**.

**monkey_patch_logging**
:   Monkeypatch **logging.Formatter** to add colors to the logs on a terminal, not customizable for now. Monkeypatch **logging.LogRecord.getMessage** to log formatting errors in the log format instead of throwing useless exception.

**logging_basic_color_config(level='DEBUG', fmt='%(asctime)s %(message)s  # %(filename)s:%(lineno)d %(name)s', color_on_terminal=True)**
:   Basic logging configuration, coloring and report formatting errors (**monkey_patch_logging**). Coloring code relies on '  # ' to determine the section to print in pale dark blue (comment).

**monkey_unpatch_logging**
:   Undo the monkeypatching done by **monkey_patch_logging** and remove coloring and error handling.

**COLOR_RESET_STR**
:   The color and attribute ANSI reset code.


## Similar projects

You may want to check out:

- 📦 [pypi/termcolor](https://pypi.python.org/pypi/termcolor) - the original project
- 📦 [delameter/termcolor](https://github.com/delameter/termcolor) - fork of the original termcolor_dg extends
- 📦 [pypi/colorama](https://pypi.org/project/colorama/) for more advanced options.


## Interesting links

- 📑 [Terminal Colors](https://github.com/termstandard/colors) discussion with examples
- 📑 [ANSI/VT100 Terminal Control Escape Sequences](https://www2.ccs.neu.edu/research/gpc/VonaUtils/vona/terminal/vtansi.htm)
- 📑 [ANSI escape code](https://en.wikipedia.org/wiki/ANSI_escape_code)
- 📑 [Color codes cheatsheet](https://delameter.github.io/termcolor)
- 📑 [MarkdownGuide - Basic Syntax](https://www.markdownguide.org/basic-syntax)


Terminal properties support
---------------------------
Assume this information is outdated.

| Terminal         |  bold   | dark  | underline |  blink  | reverse | concealed | 256 colors | 24-bit color |
|:-----------------|:-------:|:-----:|:---------:|:-------:|:-------:|:---------:|:----------:|:------------:|
| **linux**        |  ❌[^3]  | ❌[^3] |   ✅[^3]   |  ✅[^3]  |    ✅    |     ❌     |     ❌      |      ❌       |
| **konsole**      |    ✅    |   ✅   |     ✅     |    ✅    |    ✅    |     ✅     |     ✅      |      ✅       |
| **terminator**   |    ✅    |   ✅   |     ✅     |    ✅    |    ✅    |     ✅     |     ✅      |      ✅       |
| **kitty**        |    ✅    |   ✅   |     ✅     |    ✅    |    ✅    |     ❌     |     ✅      |      ✅       |
| **xterm**        |    ✅    |   ❌   |     ✅     |    ✅    |    ✅    |     ✅     |     ✅      |      ✅       |
| **rxvt**         |    ✅    |   ❌   |     ✅     |    ✅    |    ✅    |     ❌     |     ✅      |      ❌       |
| **dtterm**       |    ✅    |   ✅   |     ✅     | reverse |    ✅    |     ✅     |     ❓      |      ❓       |
| **teraterm**     | reverse |   ❌   |     ✅     | rev/red |    ✅    |     ❌     |     ❓      |      ❓       |
| **aixterm**      | normal  |   ❌   |     ✅     |    ❌    |    ✅    |     ✅     |     ❓      |      ❓       |
| **Windows**      |    ❌    |   ❌   |     ❌     |    ❌    |    ✅    |   ✅[^4]   |     ❓      |      ✅       |
| **PuTTY**        |  ✅[^2]  |   ✅   |     ✅     |  ✅[^1]  |    ✅    |     ❌     |     ✅      |      ✅       |
| **Cygwin SSH**   |    ✅    |   ❌   |   color   |  color  |  color  |     ✅     |     ❓      |      ❓       |
| **Mac Terminal** |    ✅    |   ❓   |     ✅     |    ✅    |    ✅    |     ✅     |     ✅      |      ❓       |
| **iTerm2**       |    ✅    |   ❓   |     ✅     |    ✅    |    ✅    |     ✅     |     ✅      |      ✅       |

[^1]: Disabled by default

[^2]: Supports color change, bold text or both.

[^3]: See [VGA text mode](https://en.wikipedia.org/wiki/VGA_text_mode)

[^4]: See [Add support for the "concealed" graphic rendition attribute #6876 ](https://github.com/microsoft/terminal/issues/6876)


CHANGES.md
termcolor_dg changes
====================

0.9.3.3 (2023-05-18)
--------------------

- fix DISABLED detection
- logging.basicConfig has no disable_existing_loggers argument.
- move back to GitHub, workflows does work and is free.

0.9.3 (2022-02-14)
------------------

- Adjust the log color scheme.
- pylint fixes in the color demo.

0.9.3 (2022-02-14)
------------------

- Adjust the log color scheme.
- pylint fixes in the color demo.

0.9.2 (2022-02-07)
------------------

- Combine color, background and attributes in a single escape sequence.
- Minor enhancements.

0.9.1 (2022-02-05)
------------------

- Rename to termcolor_dg (termcolor2 is taken).

0.9.0 (2022-02-05)
------------------

- Fork to termcolor2.
- Add 24-bit colors support.
- Add logging and color demos.
- Add ANSI_COLORS_DISABLED and ANSI_COLORS_FORCE environment variables.
- Add always_colored, rainbow_color.
- Add monkey_patch_logging and logging_basic_color_config utility functions.
- Better cprint and print compatibility (no arguments = new line).
- Drop 'gray' and 'on_gray' (it was an alias for black, bad idea?).
- Make the trailing reset sequence optional for colored.
- Use `python -m build` to build the project.
- Remove regex usage, the following would have not been stripped but works: `echo -e '(\e[01;32mx\e[1;0033mx\e[m)'`.
- Remove PKG-INFO.
- README.md instead of README.rst.
- ... unit tests, INSTALL ...

---

termcolor changes
=================

1.9.1 (2021-12-15)
------------------

- python 2 and 3 compatibility, avoid KeyError
- make pylint happy
- fix character escapes
- better .gitignore
- fix __ALL__ to __all__

1.1.0 (2011-01-13)
------------------

- Added cprint function.

1.0.1 (2011-01-13)
------------------

- Updated README.rst.

1.0.0 (2011-01-13)
------------------

- Changed license to MIT.
- Updated copyright.
- Refactored source code.

0.2 (2010-09-07)
------------------

- Added support for Python 3.x.

0.1.2 (2009-06-04)
------------------

- Fixed bold characters. (Thanks Tibor Fekete)

0.1.1 (2009-03-05)
------------------

- Some refactoring.
- Updated copyright.
- Fixed reset colors.
- Updated documentation.

0.1 (2008-06-09)
----------------

- Initial release.


src/termcolor2.py
#!/usr/bin/python
# -*- coding: utf-8 -*-
# Copyright (c) 2008-2011 Volvox Development Team
#
# Permission is hereby granted, free of charge, to any person obtaining a copy
# of this software and associated documentation files (the "Software"), to deal
# in the Software without restriction, including without limitation the rights
# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
# copies of the Software, and to permit persons to whom the Software is
# furnished to do so, subject to the following conditions:
#
# The above copyright notice and this permission notice shall be included in
# all copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
# THE SOFTWARE.
#
# Original author: Konstantin Lepa <konstantin.lepa@gmail.com>

'''ANSII Color formatting for output in terminal.

Run `python -m termcolor2` for color demo

Run `python -m termcolor2 logs` for colored logs demo
'''

from __future__ import absolute_import, print_function, division

import logging
import os
import sys
import time


__all__ = ['always_colored', 'colored', 'cprint', 'rainbow_color', 'monkey_patch_logging', 'logging_basic_color_config',
           'monkey_unpatch_logging', 'monkey_unpatch_logging_format']

__version__ = '0.9.0'
__copyright__ = 'Copyright (c) 2008-2011 Volvox Development Team'
__license__ = 'MIT'

__author__ = 'Konstantin Lepa'
__email__ = 'konstantin.lepa@gmail.com'

__maintainer__ = 'Doncho N. Gunchev'
__maintainer_email__ = 'dgunchev@gmail.com'

__credits__ = ['Edmund Huber', 'Lukasz Balcerzak', 'Hendrik Buschmeier', 'Nat Meysenburg', 'Iulian PAUN']


# Python 2 and 3 compatibility
if sys.version_info >= (3, 0):  # pragma: no cover
    # raw_input = input  # @ReservedAssignment pylint: disable=C0103,redefined-builtin
    # unicode = str  # @ReservedAssignment pylint: disable=C0103,redefined-builtin
    basestring = str  # @ReservedAssignment pylint: disable=C0103,redefined-builtin
    # long = int  # @ReservedAssignment pylint: disable=C0103,redefined-builtin

DISABLED = os.getenv('ANSI_COLORS_DISABLED') is not None \
    or not sys.stdout.isatty() and os.getenv('ANSI_COLORS_FORCE') is None

COLOR_RESET_STR = '\033[0m'
ERASE_EOL_STR = '\033[2K'
RESET_STR = '\033c'
FMT16_STR = '\033[%sm'
FMT256_STR = '\033[%d;5;%dm'
FMT24BIT_STR = '\033[%d;2;%d;%d;%dm'  # https://gist.github.com/lifepillar/09a44b8cf0f9397465614e622979107f

ATTRIBUTES = {
    'bold': 1,
    'dark': 2,
    'underline': 4,
    'blink': 5,
    'reverse': 7,
    'concealed': 8,
}

COLORS = {
    'black': 30,
    'red': 31,
    'green': 32,
    'yellow': 33,
    'blue': 34,
    'magenta': 35,
    'cyan': 36,
    'light_grey': 37,
    'dark_grey': 90,
    'light_red': 91,
    'light_green': 92,
    'light_yellow': 93,
    'light_blue': 94,
    'light_magenta': 95,
    'light_cyan': 96,
    'white': 97,
}

HIGHLIGHTS = {
    'black': 40,
    'red': 41,
    'green': 42,
    'yellow': 43,
    'blue': 44,
    'magenta': 45,
    'cyan': 46,
    'light_grey': 47,
    'dark_grey': 100,
    'light_red': 101,
    'light_green': 102,
    'light_yellow': 103,
    'light_blue': 104,
    'light_magenta': 105,
    'light_cyan': 106,
    'white': 107,
}


def always_colored(text, color=None, on_color=None, attrs=None, reset=True):
    '''Colorize text

    color (text color): 'black', 'red', 'green', 'yellow', 'blue', 'magenta', 'cyan', 'light_grey', 'dark_grey',
        'light_red', 'light_green', 'light_yellow', 'light_blue', 'light_magenta', 'light_cyan', 'white'.

    on_color (text background): same as color but with 'on_' or 'on ' prefix.

    attributes: 'bold', 'dark', 'underline', 'blink', 'reverse', 'concealed'.

    reset: if set to false do not emit reset sequence at the end.

    Additionally, if 256 colors are supported, any integer between 1 and 255 can be provided for both foreground
    and background. A tuple/list with three integers (R, G, B) can be provided for 24 bit color.

    Examples:
        always_colored('Hello, World!', 'red', 'on_black', ['bold', 'blink'])
        always_colored('Hello, World!', 191, 182)
        always_colored('24bit color!', (255, 127, 127), (127, 127, 255), ['bold'])
    '''
    pfx = ''

    if color is not None:
        if isinstance(color, basestring):
            if color in COLORS:
                pfx = FMT16_STR % (COLORS[color])
            else:
                raise ValueError("Invalid color %r" % color)
        elif isinstance(color, int):
            if 0 <= color <= 255:
                pfx = FMT256_STR % (38, color)
            else:
                raise ValueError("Invalid color %d" % color)
        elif isinstance(color, (list, tuple)):
            if len(color) == 3 and 0 <= color[0] <= 255 and 0 <= color[1] <= 255 and 0 <= color[2] <= 255:
                pfx = FMT24BIT_STR % (38, color[0], color[1], color[2])
            else:
                raise ValueError("Invalid color %r" % (color,))
        else:
            raise TypeError("Unsupported color type %s" % type(color).__name__)

    if on_color is not None:
        if isinstance(on_color, basestring):
            if on_color.startswith('on_'):
                on_color = on_color[3:]
            if on_color in HIGHLIGHTS:
                pfx += FMT16_STR % (HIGHLIGHTS[on_color])
            else:
                raise ValueError("Invalid on_color %r" % on_color)
        elif isinstance(on_color, int):
            if 0 <= on_color <= 255:
                pfx += FMT256_STR % (48, on_color)
            else:
                raise ValueError("Invalid on_color %d" % on_color)
        elif isinstance(on_color, (list, tuple)):

            if len(on_color) == 3 and 0 <= on_color[0] <= 255 and 0 <= on_color[1] <= 255 and 0 <= on_color[2] <= 255:
                pfx += FMT24BIT_STR % (48, on_color[0], on_color[1], on_color[2])
            else:
                raise ValueError("Invalid on_color %r" % (on_color, ))
        else:
            raise TypeError("Unsupported on_color type %s" % type(on_color).__name__)

    if attrs is not None:
        for attr in [attrs] if isinstance(attrs, basestring) else attrs:
            if attr in ATTRIBUTES:
                pfx += FMT16_STR % (ATTRIBUTES[attr])
            else:
                raise ValueError("Invalid attribute %r" % attr)

    return pfx + text + (COLOR_RESET_STR if reset and pfx else '')


def colored(text, color=None, on_color=None, attrs=None, reset=True):
    '''Colorize text if running on terminal.

    Environment variables:
        - **ANSI_COLORS_FORCE**: any value (even empty) will force colorizing the text.
        - **ANSI_COLORS_DISABLED** any value (even empty) will disable colorization. Takes precedence.

    Available text colors:
        red, green, yellow, blue, magenta, cyan, white, black, light_grey, dark_grey, light_red, light_green,
        light_yellow, light_blue, light_magenta, light_cyan.

    Available text highlights:
        on_red, on_green, on_yellow, on_blue, on_magenta, on_cyan, on_black, on_white, on_light_grey, on_dark_grey,
        on_light_red, on_light_green, on_light_yellow, on_light_blue, on_light_magenta, light_cyan.

    Additionally, if 256 colors are supported, any integer between 1 and 255 can be provided for both
    foreground and background. A tuple/list with three elements (R, G, B) can be used for 24 bit color.

    Available attributes:
        bold, dark, underline, blink, reverse, concealed.

    Examples:
        colored('Hello, World!', 'red', 'on_black', ['bold', 'blink'])
        colored('Hello, World!', 191, 182)
        colored('24bit color!', (255, 127, 127), (127, 127, 255), ['bold'])
    '''
    if DISABLED:
        return text

    return always_colored(text, color, on_color, attrs, reset)


def cprint(text='', color=None, on_color=None, attrs=None, **kwargs):
    '''
    Print colorized text.

    It accepts arguments of print function.
    '''
    print(colored(text, color, on_color, attrs), **kwargs)


def rainbow_color(n, steps, nmax=255):
    '''Calculate rainbow color'''
    if not isinstance(n, int) and isinstance(steps, int):
        raise TypeError('Arguments must be integers')
    if steps < 6:
        raise ValueError('Total must be at least 6')

    n %= steps
    progress = float(n) / steps
    r_value = max(0, min(nmax, abs(3 * nmax - 6 * progress * nmax) - nmax))

    n = (n - float(steps) / 3) % steps
    progress = float(n) / steps
    g_value = max(0, min(nmax, abs(3 * nmax - 6 * progress * nmax) - nmax))

    n = (n - float(steps) / 3) % steps
    progress = float(n) / steps
    b_value = max(0, min(nmax, abs(3 * nmax - 6 * progress * nmax) - nmax))

    return round(r_value), round(g_value), round(b_value)


def monkey_patch_logging_format():
    '''Monkey patches the logging module format error report'''
    if getattr(logging.LogRecord, 'distGetMessage', None) is not None:
        return

    logging.LogRecord.distGetMessage = logging.LogRecord.getMessage

    def print_log_record_on_error(func):
        '''Monkeypatch for logging.LogRecord.getMessage

        Credits: https://stackoverflow.com/questions/2477934/'''

        def wrap(self, *args, **kwargs):
            '''Generate wrapper function for logging.LogRecord.getMessage'''
            try:
                return func(self, *args, **kwargs)
            except Exception as exc:  # pylint: disable=broad-except
                return 'Unable to create log message with msg=%r, args=%r: %r' \
                    % (getattr(self, 'msg', '?'), getattr(self, 'args', '?'), exc)

        return wrap

    # Monkeypatch the logging library for more informative formatting errors
    logging.LogRecord.getMessage = print_log_record_on_error(logging.LogRecord.getMessage)


def monkey_unpatch_logging_format():
    '''Undo monkey_patch_logging_format'''
    if getattr(logging.LogRecord, 'distGetMessage', None) is None:
        return
    logging.LogRecord.getMessage = logging.LogRecord.distGetMessage  # @UndefinedVariable
    delattr(logging.LogRecord, 'distGetMessage')


def monkey_patch_logging(color_on_terminal=True):
    '''Monkey patches the logging module and add colors if enabled'''

    if getattr(logging, 'DistFormatter', None) is not None:
        return True

    monkey_patch_logging_format()

    if color_on_terminal and not DISABLED:
        # Monkey patches the logging module to print in color'

        def get_formatter(logging_formatter=logging.Formatter):
            '''Get it? ;-)'''

            class ColoredFormatter(logging_formatter):
                '''Color console formatter'''

                def format(self, record):
                    output = logging_formatter.format(self, record)
                    tail = None
                    comment_pos = output.find('  # ')  # Intentionally require two spaces before
                    if comment_pos >= 0:
                        output, tail = output[:comment_pos], output[comment_pos:]

                    if record.levelno < logging.DEBUG:  # pylint: disable=fixme
                        output = colored(text=output, color='black', on_color='on_blue', attrs=['dark'])
                    elif record.levelno <= logging.DEBUG:  # pylint: disable=fixme
                        output = colored(text=output, color='black', on_color=None, attrs=['dark'])
                    elif record.levelno <= logging.INFO:
                        output = colored(text=output, color='green', on_color=None, attrs=['bold'])
                    elif record.levelno <= logging.WARNING:
                        output = colored(text=output, color='yellow', on_color=None, attrs=['bold'])
                    elif record.levelno <= logging.ERROR:
                        output = colored(text=output, color='red', on_color=None, attrs=['bold'])
                    elif record.levelno <= logging.CRITICAL:
                        output = colored(text=output, color='white', on_color='on_red', attrs=['bold'])
                    else:
                        output = colored(text=output, color='yellow', on_color='on_red',
                                         attrs=['bold', 'blink', 'underline'])
                    if tail:
                        output += colored(text=tail, color='blue', attrs=['dark'])

                    return output

                def formatException(self, ei):
                    '''Format and return the specified exception information as a string.'''
                    text = logging_formatter.formatException(self, ei)
                    text = '\n'.join(ERASE_EOL_STR + i for i in text.splitlines())
                    return colored(text=text, color='white', on_color='on_magenta', attrs=('bold'))

            return ColoredFormatter

        logging.DistFormatter = logging.Formatter
        logging.Formatter = get_formatter()

    return True


def monkey_unpatch_logging():
    '''Undo monkey_patch_logging'''

    monkey_unpatch_logging_format()

    if getattr(logging, 'DistFormatter', None) is not None:
        logging.Formatter = logging.DistFormatter  # @UndefinedVariable
        delattr(logging, 'DistFormatter')


def logging_basic_color_config(level='DEBUG', fmt='%(asctime)s %(message)s  # %(filename)s:%(lineno)d %(name)s',
                               color_on_terminal=True):
    '''Setup basic logging with fancy format and colors if running on terminal

    a very fancy fmt would be "%(asctime)s %(levelname)-8s: %(message)s  # %(filename)s:%(lineno)d %(name)s"
    '''
    monkey_patch_logging(color_on_terminal=color_on_terminal)
    try:
        logging.basicConfig(level=level, format=fmt, disable_existing_loggers=False)
    except ValueError:
        logging.basicConfig(level=level, format=fmt)


def termcolor_demo_16():
    '''Base 16 color demo'''
    colors = ('black', 'red', 'green', 'yellow', 'blue', 'magenta', 'cyan', 'light_grey', 'dark_grey', 'light_red',
              'light_green', 'light_yellow', 'light_blue', 'light_magenta', 'light_cyan', 'white')
    max_len = max(len(color) for color in colors)

    print(RESET_STR, end='')
    print(('--- 16 color mode test on TERM=%r' % os.getenv('TERM')).ljust(119, '-'))

    for i, color in enumerate(colors):
        print(' ', end='')
        cprint(color.replace('_', ' ').center(max_len), color=color,
               on_color='on_black' if color != 'black' else 'on_dark_grey', end='' if i != 7 else '\n')
    print()

    for i, color in enumerate(colors):
        print(' ', end='')
        cprint(color.replace('_', ' ').center(max_len), color='black' if color != 'black' else 'dark_grey',
               on_color='on_' + color, end='' if i != 7 else '\n')
    print()

    print('Attributes:')
    all_attrs = ('bold', 'dark', 'underline', 'blink', 'reverse', 'concealed')
    attr_max_len = max(len(attr) for attr in all_attrs)

    print(' [' + colored('None'.center(attr_max_len), color='light_green', on_color='on_black') + ']', end='')
    for attr in all_attrs:
        print(' [' + colored(attr.center(attr_max_len), color='light_green', on_color='on_black', attrs=[attr]) + ']',
              end='')
    print(' last one should be invisible')

    print()
    print(' [' + '   Bold black on black    ', end=']')
    print(' [' + '       Dark white         ', end=']')
    print(' [' + '     Underline green      ', end=']')
    print(' [' + '      Blink yellow        ', end=']\n')
    print(' [' + colored('   Bold black on black    ', 'black',   'on_black',   attrs=['bold']), end=']')
    print(' [' + colored('       Dark white         ', 'white',   attrs=['dark']), end=']')
    print(' [' + colored('     Underline green      ', 'green',   attrs=['underline']), end=']')
    print(' [' + colored('      Blink yellow        ', 'yellow',  attrs=['blink']), end=']\n')

    print(' [' + colored('      Reversed blue       ', 'blue',    attrs=['reverse']), end=']')
    print(' [' + colored('    Concealed Magenta     ', 'magenta', attrs=['concealed']), end=']')
    print(' [' + colored('Bold underline reverse red', 'red',     attrs=['bold', 'underline', 'reverse']), end=']')
    print(' [' + colored('Dark blink concealed white', 'white',   attrs=['dark', 'blink', 'concealed']), end=']\n')
    print(' [' + '      Reversed blue       ', end=']')
    print(' [' + '    Concealed Magenta     ', end=']')
    print(' [' + 'Bold underline reverse red', end=']')
    print(' [' + 'Dark blink concealed white', end=']\n')

    def all_combos(values):
        values = tuple(values)
        if values:
            for i in all_combos(values[1:]):
                yield i
                yield [values[0]] + i
        else:
            yield []

    print('\nAttribute salad (Bold, Dark, Underline, blinK, Reverse, Concealed):')
    for idx, attrs in enumerate(all_combos(all_attrs)):
        code = 'B' if 'bold' in attrs else 'b'
        code += 'D' if 'dark' in attrs else 'd'
        code += 'U' if 'underline' in attrs else 'u'
        code += 'K' if 'blink' in attrs else 'k'
        code += 'R' if 'reverse' in attrs else 'r'
        code += 'C' if 'concealed' in attrs else 'c'
        print(' [' if idx % 16 == 0 else ' ' + colored(code, attrs=attrs), end='' if idx % 16 != 15 else ' ]\n')


def termcolor_demo_256():
    '''256 color demo'''
    print('\n--- 256 color mode test '.ljust(120, '-'))
    print(' First 16: [', end='')
    for i in range(16):
        cprint('%3d ' % i, color=15 - i, on_color=i, end='')
    print(']')

    print()

    print('6*6*6 cube:')
    for y in range(6):
        for x in range(6):
            print(' ', end='')
            for i in range(6):
                background = 16 + y * 6 * 6 + x * 6 + i
                cprint('%3x' % background, color='white' if y < 3 else 'black', on_color=background, end='')
        print()

    print()

    print(' 24 grayscale test:  [', end='')
    for i in range(232, 256):
        # cprint('%4d' % i, color='white' if i < 244 else 'black', on_color=i, end='')
        cprint('%4d' % i, color=(232 + 255) - i, on_color=i, end='')
    print(']\n')


def termcolor_demo_24bit():
    '''24 bit color demo'''
    data = '=== 24 bit color mode test '.ljust(114, '=')
    for i in range(2 * len(data), -1, -1):
        print('\r   ', end='')
        grey = abs(i * 8 % 510 - 255)
        print(
            colored(
                ''.join(colored(char, on_color=rainbow_color(step + i, len(data)), reset=False)
                        for step, char in enumerate(data)),
                color=(grey, grey, grey)
            ), end='')

        sys.stdout.flush()
        time.sleep(0.02)

    print('\r   ', end='')
    for step, char in enumerate(data):
        print(colored(char,
                      rainbow_color(step - len(data) // 3 + i, len(data)),
                      on_color=rainbow_color(step + i, len(data)), reset=False), end='')
        sys.stdout.flush()
        time.sleep(0.02)
        print(COLOR_RESET_STR, end='')

    print()


def termcolor_demo():
    '''Demonstrate this module's capabilities'''
    termcolor_demo_16()
    termcolor_demo_256()
    termcolor_demo_24bit()


def color_log_demo():
    '''Test color logging on terminal and logging format error'''
    # monkey_patch_logging(color_on_terminal=True)
    logging_basic_color_config()

    log = logging.getLogger('logger')
    print('Logging test... levels and exception:')

    # Hack to skip the log level test
    log._log(logging.NOTSET, 'Not set, below DEBUG, normally not show...', [])  # pylint: disable=W0212
    log.debug('Debug')
    log.info('Info')
    log.warning('Warning')
    log.error('Error')
    log.critical('Critical')
    log.error('', 1)  # the mistake is intentional pylint: disable=logging-too-many-args
    log.info('%d', 1, 2)  # the mistake is intentional pylint: disable=logging-too-many-args
    log.debug('%d %d', 'a', 2)
    print()
    try:
        raise TypeError('msg')
    except TypeError:
        log.exception('Exception')
    print()
    log.log(51, 'ABOVE CRITICAL! intentionally annoying, we should never see this IRL...')
    print()
    log.info('Done.')


def main():
    '''Main demo entry point, if no arguments - color demo, if any - colored logs demo'''
    if len(sys.argv) == 1:       # pragma: no cover
        return termcolor_demo()  # pragma: no cover
    return color_log_demo()      # pragma: no cover


if __name__ == '__main__':
    sys.exit(main())
